internal/manager/telemetry/telemetry_test.go

```diff
@@ -0,0 +1,430 @@
+	"k8s.io/apimachinery/pkg/version"
+	discoveryclient "k8s.io/client-go/discovery"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+	gatewayv1alpha2 "sigs.k8s.io/gateway-api/apis/v1alpha2"
+	gatewayv1beta1 "sigs.k8s.io/gateway-api/apis/v1beta1"
+	gatewayclient "sigs.k8s.io/gateway-api/pkg/client/clientset/versioned"
+
+	"github.com/kong/kubernetes-ingress-controller/v2/internal/manager"
+	"github.com/kong/kubernetes-ingress-controller/v2/internal/manager/featuregates"
+	"github.com/kong/kubernetes-ingress-controller/v2/internal/manager/telemetry"
+	"github.com/kong/kubernetes-ingress-controller/v2/internal/util"
+	"github.com/kong/kubernetes-ingress-controller/v2/pkg/clientset/scheme"
+	"github.com/kong/kubernetes-ingress-controller/v2/test/envtest"
+)
+
+func TestTelemetry(t *testing.T) {
+	t.Log("configuring TLS listener - server for telemetry data")
+	cert, err := generateSelfSignedCert()
+	require.NoError(t, err)
+	listener, err := tls.Listen("tcp", "localhost:0", &tls.Config{
+		Certificates: []tls.Certificate{cert},
+		// The same version as the one used by TLS forwarder in the pkg telemetry.
+		MinVersion: tls.VersionTLS13,
+		MaxVersion: tls.VersionTLS13,
+	})
+	require.NoError(t, err)
+	defer listener.Close()
+	// Run a server that will receive the report, it's expected
+	// to be the first connection and the payload.
+	reportChan := make(chan []byte)
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	go func() {
+		// Any function return indicates that either the
+		// report was sent or there was nothing to send.
+		defer close(reportChan)
+		conn, err := listener.Accept()
+		if !assert.NoError(t, err) {
+			return
+		}
+		defer conn.Close()
+		for {
+			report := make([]byte, 2048) // Report is much shorter.
+			n, err := conn.Read(report)
+			if !assert.NoError(t, err) {
+				return
+			}
+			select {
+			case reportChan <- report[:n]:
+			case <-ctx.Done():
+				return
+			}
+		}
+	}()
+
+	t.Log("configuring envtest and creating K8s objects for telemetry test")
+	envcfg := envtest.Setup(t, scheme.Scheme)
+	cfg := configForEnvTestTelemetry(t, envcfg)
+	c, err := cfg.GetKubeconfig()
+	require.NoError(t, err)
+	createK8sObjectsForTelemetryTest(ctx, t, c)
+
+	t.Log("starting the controller manager")
+	// Override the telemetry settings, to allow testing.
+	// Set them back to the original values at the end of the test.
+	set := func(ep string, skipVerify bool, dur time.Duration) {
+		telemetry.SplunkEndpoint = ep
+		telemetry.SplunkEndpointInsecureSkipVerify = skipVerify
+		telemetry.TelemetryPeriod = dur
+	}
+	defer set(telemetry.SplunkEndpoint, telemetry.SplunkEndpointInsecureSkipVerify, telemetry.TelemetryPeriod)
+	set(listener.Addr().String(), true, time.Second) // Let's have long duration due too rate limiter in K8s client.
+	go func() {
+		deprecatedLogger, _, err := manager.SetupLoggers(&cfg, io.Discard)
+		if !assert.NoError(t, err) {
+			return
+		}
```

Comment:
```suggestion
		require.NoError(t, err)
```

but I'm not sure if that wouldn't panic if `err != nil` in this goroutine 🤔